File: frontend/types/transaction.ts
import { SyncStatus } from "./sync";

export type Transaction = {
  id: string;
  type: 'expense' | 'income';
  merchant?: string | null;  // merchant can be null
  place: string;
  amount: number;
  category: string;
  account: string;
  tags?: string; // JSON array of tag IDs
  date: string;  // Always ISO string for consistency
  description: string;
  uid: string;
  sync_status: SyncStatus;
  deleted_at?: number;  // Unix timestamp in milliseconds
  created_at?: number;  // Unix timestamp in milliseconds
  updated_at: number;   // Unix timestamp in milliseconds
};

export type TransactionListType = {
  data: Transaction[];
};

export type TransactionItemProps = {
  item: Transaction;
  index: number;
  handleClick: Function;
};

File: backend/app/models/transaction.py
from pydantic import BaseModel, Field
from typing import Literal, Optional

class TransactionBase(BaseModel):
    type: Literal['expense', 'income']
    amount: float = Field(..., gt=0)
    category: str = Field(..., min_length=1)
    date: str  # ISO date string (YYYY-MM-DD)
    description: str = Field(default="")

class TransactionCreate(TransactionBase):
    uid: str
    id: str

class TransactionResponse(TransactionBase):
    id: str
    uid: str
    created_at: int  # Unix timestamp in milliseconds
    updated_at: int  # Unix timestamp in milliseconds
    deleted_at: Optional[int] = None  # Unix timestamp in milliseconds